### ISSUE TITLE
Adding Custom Classes for Grunt Mod

### ISSUE THREAD

**Issue by sbmmsux** [DBmaoha/GruntMode#1]
Hello there. How can we set up custom classes for this mod? In the Northstar website you said something about using a few commands but, where and how? Can we set classes in the "private match" menu and then use them in your mod? Or, do we have to set the classes manually in the codelines of your mod? Including attachments in the weapons.. Please help.. And also, is there any way to use this mod in a different game mode? I mean, not only "Attrition". Thanks.

**Comment by DBmaoha** [comment]
this version is outdated, for newest and stable version, please check out https://github.com/DBmaoha/Super.Mixed.Game( haven't translated to english though )

1. How can we set up custom classes for this mod?
You can set up a class by using the function `GruntMode_RegisterPlayerClass()` in any scripts.
Parameters:
-className: your new class' name, cannot be the same as any other classes
-loadoutFunction: how this class is given weapons. classMods and so on
-isSpecialist: whether killing this class will get extra scores
I recommand copy & paste the function `SimpleClassLoadout()` from `_grunt_default_classes.gnut`, you can also learn how to create a class from this file.

2. Is there any way to use this mod in a different game mode?
There is, but only attrition has the popups like "Score +xx" and it's scoreboard shows the status you kill npcs, players and titans, which is specified for ai combats, so I perfer not to change gamemode.

**Comment by sbmmsux** [comment]
Thanks.. Well, I tried that other version you recommended me and... things didnt go so well.. I got this error message.. 
![2](https://user-images.githubusercontent.com/130800146/232327794-341a2d09-9eb8-4c9f-ae33-72bebec7ec47.png)

Any idea of why it is happening? And other thing.. I know you already answered me how to register my own custom classes in the mod but.. To be honest, it was like reading chinese jejeje.. I know shit about programming.. Let's say I wanna make a class with a R97 and an ACOG sight (or similar) as an attachment, how would I have to write that? I guess it is in the "defaultclasses" file, right? That's all I understand so far..

**Comment by sbmmsux** [comment]
Funny thing.. I put the old GruntMod version back and, the game started normal.. no error or anything..

**Comment by sbmmsux** [comment]
Forget it.. now the error happens with any version of your mod.. You can log in properly only by disabling the Grunt Mod.. so, what's going on? I saw that u uploaded sevaral updates but.. same outcome with any of those..

**Comment by DBmaoha** [comment]
> Forget it.. now the error happens with any version of your mod.. You can log in properly only by disabling the Grunt Mod.. so, what's going on? I saw that u uploaded sevaral updates but.. same outcome with any of those..

Error message on your console please, these are not enough.

**Comment by DBmaoha** [comment]
> Thanks.. Well, I tried that other version you recommended me and... things didnt go so well.. I got this error message.. ![2](https://user-images.githubusercontent.com/130800146/232327794-341a2d09-9eb8-4c9f-ae33-72bebec7ec47.png)
> 
> Any idea of why it is happening? And other thing.. I know you already answered me how to register my own custom classes in the mod but.. To be honest, it was like reading chinese jejeje.. I know shit about programming.. Let's say I wanna make a class with a R97 and an ACOG sight (or similar) as an attachment, how would I have to write that? I guess it is in the "defaultclasses" file, right? That's all I understand so far..

Function name has been updated, maybe you should check it

**Comment by DBmaoha** [comment]
> Thanks.. Well, I tried that other version you recommended me and... things didnt go so well.. I got this error message.. ![2](https://user-images.githubusercontent.com/130800146/232327794-341a2d09-9eb8-4c9f-ae33-72bebec7ec47.png)
> 
> Any idea of why it is happening? And other thing.. I know you already answered me how to register my own custom classes in the mod but.. To be honest, it was like reading chinese jejeje.. I know shit about programming.. Let's say I wanna make a class with a R97 and an ACOG sight (or similar) as an attachment, how would I have to write that? I guess it is in the "defaultclasses" file, right? That's all I understand so far..

The function `RegisterGruntModePlayerClass()` has been updated to `GruntMode_RegisterPlayerClass()`, maybe you should check it

**Comment by DBmaoha** [comment]
if you want to use the repository directly, go https://github.com/DBmaoha/Super.Mixed.Game/tree/english-gruntmode but not the main branch, the main branch is for northstarCN witch is not translated to english and has some specified functions

**Comment by DBmaoha** [comment]
> Thanks.. Well, I tried that other version you recommended me and... things didnt go so well.. I got this error message.. ![2](https://user-images.githubusercontent.com/130800146/232327794-341a2d09-9eb8-4c9f-ae33-72bebec7ec47.png)
> 
> Any idea of why it is happening? And other thing.. I know you already answered me how to register my own custom classes in the mod but.. To be honest, it was like reading chinese jejeje.. I know shit about programming.. Let's say I wanna make a class with a R97 and an ACOG sight (or similar) as an attachment, how would I have to write that? I guess it is in the "defaultclasses" file, right? That's all I understand so far..

Register a new class, to give them a r97 with attachment( r97 don't have acog unfortunately, maybe use holosight? ), leave `SimpleClassLoadout()`'s second and third parameters empty, then use the method `player.GiveWeapon()`

```
void function YourFunctionName( entity player )
{
	player.SetPlayerSettings( "pilot_grapple_male" ) // have a better pov model

	SimpleClassLoadout(  
		player,
		"",																	// primary weapon, leave it empty to set them up yourself
		"",																	// secondary weapon, leave it empty to set them up yourself
		"mp_weapon_frag_grenade",											// grenade
		$"models/humans/grunts/imc_grunt_smg.mdl",							// imc player model
		$"models/humans/grunts/mlt_grunt_smg.mdl",							// mlt player model
		["disable_doublejump", "only_wallclimb"],							// class mods
		"Class's Title",													// title
		"Class's respawn notification"										// respawn notification
	)

	player.GiveWeapon( "mp_weapon_r97", ["holosight"] ) // r97 with holosight
	player.GiveWeapon( "mp_weapon_rocket_launcher" ) // archer
}
```

**Comment by sbmmsux** [comment]
Alright, I think I got it this time.. thanks.. and yeah, it aint have an acog sight.. just an example.. To be honest I barely know the weapons from Titanfall, played it like.. 5 years ago and only the campaign. Anyway, I'll try the version of the mod you suggested.. that "server compilation" error came outta nowhere.. I had the v1.7 of your mod and all good.. then updated to the 2.0.5 and everything started to fail.. and now even with the v1.7 the error shows up.. crazy..

**Comment by sbmmsux** [comment]
Quick update.. it keeps showing the same error.. "server script compilation"..

**Comment by DBmaoha** [comment]
> Quick update.. it keeps showing the same error.. "server script compilation"..

Press "~" to check your console, it lists exact the error happens

**Comment by sbmmsux** [comment]
sure.. here.. this is what showed up.. 
![4](https://user-images.githubusercontent.com/130800146/232395295-8323c7fd-afe8-40b4-ad5a-3eb34ff47967.png)

**Comment by DBmaoha** [comment]
> sure.. here.. this is what showed up.. ![4](https://user-images.githubusercontent.com/130800146/232395295-8323c7fd-afe8-40b4-ad5a-3eb34ff47967.png)

my bad, forget that it requires Karma.Abuse

**Comment by DBmaoha** [comment]
newest version should be good

**Comment by sbmmsux** [comment]
Oh cool.. that was quick.. I'll try it and let ya know..

**Comment by DBmaoha** [comment]
> Oh cool.. that was quick.. I'll try it and let ya know..

I uploaded v2.1.2 but it's still not show up on thunderstore, using this https://github.com/DBmaoha/Super.Mixed.Game/tree/english-gruntmode should also be good

**Comment by sbmmsux** [comment]
Yes sirr!! The mod's working now.. Thanks.. Actually now I see more action in the battlefield.. some npc's that werent there before: ticks, other type of grunts, etc.. Anyway.. only two questions after testing this version: 1) what happened to the announcer? now I aint hear her/him anymore.. the rest of the npc's are okay telling their lines.. and 2) Is there a way to disable that constant map loading? I mean, when the match ends, the mod immediately loads a new map for another match.. reminds me to black ops 1 when u only wanted to play a casual match against bots, and then the game loaded a new map for another match just when u didnt want to play anymore jeje..

**Comment by DBmaoha** [comment]
> Yes sirr!! The mod's working now.. Thanks.. Actually now I see more action in the battlefield.. some npc's that werent there before: ticks, other type of grunts, etc.. Anyway.. only two questions after testing this version: 1) what happened to the announcer? now I aint hear her/him anymore.. the rest of the npc's are okay telling their lines.. and 2) Is there a way to disable that constant map loading? I mean, when the match ends, the mod immediately loads a new map for another match.. reminds me to black ops 1 when u only wanted to play a casual match against bots, and then the game loaded a new map for another match just when u didnt want to play anymore jeje..

the announcer has been disabled, I added a function `SetShouldPlayFactionDialogue( bool enable )` for controlling them. to enable again, you can go `becomesgrunt.gnut` and change `SetShouldPlayFactionDialogue( false )` to true.
if you don't like the instant map change, you can go `gruntmode.gnut` and comment out the `AddCallback_GameStateEnter( eGameState.Postmatch, OnGamePostmatch )`.

**Comment by DBmaoha** [comment]
I'll add a playlistvar for the map change, you can use it later

**Comment by sbmmsux** [comment]
Yup, I managed to do all of that before u answered by trial and error, also deactivated that line which turned the maps into night time.. thanks anyway. Wanted to ask ya.. ur mod aint work with Northstar 1.14, right? I had to let mine in 1.12.7 version.. otherwise the GruntMode shows an error message.. But it's fine, I mean.. that 1.14 version only matters coz it allows ya to play BountyHunt mode but.. as a PvE mode.. that mode's not the best choice. Is there any way to play "Frontier War" mode on a private match? That mode looks great, like a mix between Domination and Attrition.. tried to set it up but, the option is disabled.

**Comment by DBmaoha** [comment]
> Yup, I managed to do all of that before u answered by trial and error, also deactivated that line which turned the maps into night time.. thanks anyway. Wanted to ask ya.. ur mod aint work with Northstar 1.14, right? I had to let mine in 1.12.7 version.. otherwise the GruntMode shows an error message.. But it's fine, I mean.. that 1.14 version only matters coz it allows ya to play BountyHunt mode but.. as a PvE mode.. that mode's not the best choice. Is there any way to play "Frontier War" mode on a private match? That mode looks great, like a mix between Domination and Attrition.. tried to set it up but, the option is disabled.

Sorry for my late quote after so many days, nice work figuring out youself!
The "Frontier War" and "Bounty Hunt" are all implemented by me, newest version of the SuperMixedGame repository should be compatible with northstar v1.14, you can update both
To play Frontier War, you need northstar version of at least v1.13

**Comment by sbmmsux** [comment]
oh cool.. and where can i download that "super mixed" mod? it's just that.. i download the mods manually in the northstar website.. that version is in there? all i saw was ur "grunt mode".. which is currently on the 2.5.1 version.. is that the "super mixed" ur talking about?

**Comment by DBmaoha** [comment]
> oh cool.. and where can i download that "super mixed" mod? it's just that.. i download the mods manually in the northstar website.. that version is in there? all i saw was ur "grunt mode".. which is currently on the 2.5.1 version.. is that the "super mixed" ur talking about?

https://github.com/DBmaoha/Super.Mixed.Game/tree/english-gruntmode

**Comment by sbmmsux** [comment]
Yeah, I tested.. it works with the 1.14 version of Northstar but.. so far, I didnt see the option for playing "frontier war" on a private match.. only in the servers and.. needless to say that they lag too much.. so, bottom line.. is there no way you can play "frontier war" on a private match by yourself without servers? i also managed to play "bountyhunt" mode but.. mehh..

**Comment by DBmaoha** [comment]
> Yeah, I tested.. it works with the 1.14 version of Northstar but.. so far, I didnt see the option for playing "frontier war" on a private match.. only in the servers and.. needless to say that they lag too much.. so, bottom line.. is there no way you can play "frontier war" on a private match by yourself without servers? i also managed to play "bountyhunt" mode but.. mehh..

of course you can play them in private match. just go to second page of mode selection menu, other northstar modified gamemodes are in second and third page, too

**Comment by sbmmsux** [comment]
Yeah.. I saw them finally.. silly me jejeje.. started playing Frontier War but.. once the npc's killed me the first time.. this error showed up.. 
![Sin título](https://github.com/DBmaoha/GruntMode/assets/130800146/4c5b46dd-0ece-4373-b643-560ae6e4ed91)
and off course, the match got closed.. any idea of why that happened?

**Comment by sbmmsux** [comment]
ah cool..  is it in the same link you gave me before? it's just that ur page aint have a "releases" tab so.. i cannot tell which version is which..

**Comment by DBmaoha** [comment]
> ah cool.. is it in the same link you gave me before? it's just that ur page aint have a "releases" tab so.. i cannot tell which version is which..

yes, updates will always be in that link. you can see version date here
![image](https://github.com/DBmaoha/GruntMode/assets/56738369/5e7eecbc-5e3a-4006-86c8-c297aa04a885)

**Comment by sbmmsux** [comment]
oh yeah.. saw it.. thanks..

**Comment by sbmmsux** [comment]
yup, it's all working now, it even has the epilogue working..  is there a way to implement that grunt mode of yours into that frontier war game mode? so there are a few npc's on my team..

**Comment by DBmaoha** [comment]
> yup, it's all working now, it even has the epilogue working.. is there a way to implement that grunt mode of yours into that frontier war game mode? so there are a few npc's on my team..

it is possible, but you may have to comment out `InitCampTracker( info_target )` in the function `InitFWCampSites()`, then read through frontier war's `FWAiCampThink()` function and edit it

**Comment by DBmaoha** [comment]
mostly you can try to replace it's spawner function with grunt_mode `SpawnerExtend()`'s content

**Comment by sbmmsux** [comment]
oh.. good.. and.. by the way, have you managed to make the gunships attack in that mode.. i see them showing up but.. most of the time they just stand still..

**Comment by sbmmsux** [comment]
also.. do you know how to stop northstar from updating itself? i returned to version 1.12.27 so i could use the grunt mode 2.1.2.. the bots in there are way more competitive.. in ur latest version, bots wont kill too much, for some reason.. and now i'm noticing norsthar is updating itself without me telling it to do it.

**Comment by sbmmsux** [comment]
Hey there! Dude.. I just saw the new mod u made.. it friggin rocks! Attrition but this time u can play as a pilot instead of a grunt.. cool.. Now, this is a full PvE mode what you just made.. Thanks, pal! There's another dude who's making another Grunt Mode but implementing in the "Amped Hardpoint" as well.. At least that's what he's trying. Anyway, thx for the mod!
